_posts/2023-02-21-using-nlp-for-compliance-part-2.md

```diff
@@ -0,0 +1,77 @@
+
+An added benefit of having this continual data going into Kapiche, is we can jump in and answer any high level questions on numbers and segments across all our microservices. Kapiche is very good at slicing and dicing the data, so we can generally answer any adhoc question of the data. 
+
+Here is one of our microservices ![Continous Integration, Continuous Deployment and Continous Attestation flows](/images/CI-SBOM-Analysis.png)
+
+Here is our ![DevSecOps ToolChain](/images/DevSecOps-Toolchain.jpg)
+
+
+#### Future Work
+
+**Greater Context**
+
+We are also looking for ways to add more metadata or context to this story. We use Dependency Track, which we also upload our SBOM data to, this has more vulnerability datasources, so can often give more context. It's also a lot easier if we are given an SBOM from a vendor to upload and assess it, not that we are at that stage yet. We also have our policies in there around licenses and vulnerabilities much like we have in BigQuery. It tends to do a better job of component analysis or per service analysis, so we see a much more fine grained view. I have to find time to figure out how to export the information I want into the right format in BigQuery. This should give greater context to our analysis, then we can Share our results. I also want to share more of scripts we have done, even though most of what we have used in opensource, the key bits are the joining bits.
+
+**Machine Learning**
+
+We are looking to apply more software engineering and supply chain management principles to our Machine Learning (ML) images and workflows. We currently build container images from the ML models we use, this means we download from HuggingFace or our ML Model Registry at build time instead of downloading at runtime. Three benefits we get from this, we are signing and being very declarative on what ML model and version we are using and this metadata gets included in our pipeline for analysis, the third one is faster start up times for our ML services, as sometimes these models can get up to 8GB in size. We also get the added bonuses of adding signing and provenance to our ML models, which is lacking severely and is a huge hole in the industry at the moment, as most models people use are either in cloud buckets or on Huggingface. Machine learning models should be built like other artifacts and using principles that are developing in the software supply chain like signing, provenance, attestation. The models should be then stored in artifact stores.
```

Comment:
```suggestion
We are looking to apply more software engineering and supply chain management principles to our Machine Learning (ML) images and workflows. We currently build container images from the ML models we use. This means we download from HuggingFace or our ML Model Registry at build time instead of downloading at runtime. Three benefits we get from this: we are signing and being very declarative on what ML model and version we are using; and this metadata gets included in our pipeline for analysis; and the third one is faster start up times for our ML services, as sometimes these models can get up to 8 GB in size. We also get the added bonuses of adding signing and provenance to our ML models, which is lacking severely and is a huge hole in the industry at the moment: most models people use are either in cloud buckets or on Huggingface. Machine learning models should be built like other artifacts and using principles that are developing in the software supply chain like signing, provenance, attestation. The models should be then stored in artifact stores.
```